### ISSUE TITLE
Loading from other sources than standard filesystem

### ISSUE THREAD

**Issue by yamashi** [libpd/libpd#300]
Hi,

I am using Unreal Engine 4 and files are all packed in the engine's archives. I can access the content of the .pd file but libpd_openfile won't work, is there a way to load from a buffer directly ?

Best,
Max

**Comment by danomatika** [comment]
Not as yet, although this has been requested for similar grounds (use within a game engine). It may be possible to do with some direct usage of the Pd core api, but also may require changes to the core (wrapped by libpd). Check the pd-list  and/or pd-dev archives, I don't remember where the discussion was now.

**Comment by danomatika** [comment]
Keeping this open for now.

**Comment by Spacechild1** [comment]
You can load pd patches from buffers with `binbuf_text` + `binbuf_eval`. Check the source code of `binbuf_evalfile`. You just need to replace the call to `binbuf_read` with a direct call to `binbuf_text`.

The tricky thing with virtual file systems is how to handle abstraction loading, but I think you can achieve this by registering a custom loader with `sys_register_loader`.

**Comment by danomatika** [comment]
@Spacechild1 Would you have any example code? I could try wrapping that as a libpd api call.

**Comment by Spacechild1** [comment]
Here's a `binbuf_evaltext` function: [binbuf_evaltext.zip](https://github.com/libpd/libpd/files/4777108/binbuf_evaltext.zip)

The custom loader would depend very much on the host application and consists basically of two things:
1) a way to look for abstractions by name and (fake) directory
2) a custom creator method for abstractions

When an abstraction is loaded the first time, it is registered like this:
```
class_new(gensym(objectname),
                        (t_newmethod)my_create_abstraction, 0,
                        0, 0, A_GIMME, 0)))
```
`my_create_abstraction` could look like this:
```
static t_pd *my_create_abstraction(t_symbol*s, int argc, t_atom *argv)
{
    if (!pd_setloadingabstraction(s))
    {
        t_pd *was = s__X.s_thing;
        const char *text = /* figure out how to get this */
        size_t size = /* figure out how to get this */
        t_symbol *dir = /* figure out how to get this */
        canvas_setargs(argc, argv);
        binbuf_evaltext(text, size, s, dir);
        if (s__X.s_thing && was != s__X.s_thing)
            canvas_popabstraction((t_canvas *)(s__X.s_thing));
        else s__X.s_thing = was;
        
        canvas_setargs(0, 0);
        return (pd_this->pd_newest);
    }
    else error("%s: can't load abstraction within itself\n", s->s_name);
    pd_this->pd_newest = 0;
    return (0);
}
```
Both the loader function and `my_create_abstraction` need some way to find a patch in the virtual file system. The user can decide about search paths and search order.

Also note that `pd_setloadingabstraction`, unlike `pd_pushsym` and `pd_popsym`, is not part of the public API. That's something we could change in the core.

**Comment by Spacechild1** [comment]
On the other hand, for `[declare]` to work, we need to walk the canvas environments. Unfortunately, `canvas_open` etc. only work with "real" file systems.

We could provide global hooks to replace the standard `sys_open`, `sys_close`, `read`, `write` syscalls with custom functions. This would also be needed to for other things, like reading/writing soundfiles.

**Comment by Spacechild1** [comment]
Of course, with such file system hooks you wouldn't need any of the things I've described above. Everything should just work out of the box :-) I think it's not even difficult to implement in the core.

**Comment by yamashi** [comment]
Adding that lseek needs to be overriden as well, I will look into this method, thanks !

**Comment by yamashi** [comment]
For anyone who needs something similar in the future : https://github.com/tiltedphoques/libpd/commit/e68a9f645e0235e01e2f36e852b369342fc0676a

Note that I haven't tested with anything more than just .pd files with no dependencies (which probably don't work atm)

**Comment by danomatika** [comment]
@Spacechild1 As you suggested, that looks simple enough.

**Comment by Spacechild1** [comment]
cool. you might consider making a PR to pure-data.

**Comment by Spacechild1** [comment]
I will try to implement virtual filesystem hooks for Pd. The changes should be minimal and non-invasive, so I see a good chance that Miller would include something like this. It would be really cool if Pd could work without any "real" filesystem.

**Comment by yamashi** [comment]
I considered making a PR but this is a "hack" that has limitations so I didn't think it was a good idea until those limitations are fixed.

**Comment by Spacechild1** [comment]
it has limitations in a sense that it doesn't give you full support for virtual filesystems (because it doesn't handle abstraction loading and search paths), but I think those are useful utility function, e. g. for programatically creating and opening Pd patches.
